05_concurrency/homework/main.go

```diff
@@ -0,0 +1,130 @@
+package main
+
+import (
+	"sync"
+	"time"
+)
+
+type cache interface {
+	Set(key string, value interface{}, ttl time.Duration)
+	Get(key string) (interface{}, bool)
+	Delete(key string)
+}
+
+type item struct {
+	Value      interface{}
+	Expiration int64
+	TTL        time.Duration
+}
+
+type memoryCache struct {
+	sync.Mutex
+	defaultExpirationTime time.Duration
+	cleanUpInterval       time.Duration
+	items                 map[string]item
+}
+
+func createCache(defaultExpirationTime, cleanUpInterval time.Duration) *memoryCache {
+	items := make(map[string]item)
+
+	cache := memoryCache{
+		items:                 items,
+		defaultExpirationTime: defaultExpirationTime,
+		cleanUpInterval:       cleanUpInterval,
+	}
+
+	if defaultExpirationTime <= 0 {
+		defaultExpirationTime = time.Second
+	}
+
+	if cleanUpInterval > 0 {
+		cache.startCleanUp()
+	}
+
+	return &cache
+}
+
+func (m *memoryCache) set(key string, value interface{}, ttl time.Duration) {
```

Comment:
Type `*memoryCache` does not implement `Cache` interface because its methods are unexported but interface requires them to be exported.